### PR TITLE
[HELM] Fix Kubernetes Routing Issue

### PR DESCRIPTION
**Pull request by piby180**
This PR aims to partially solve the issue mentioned [here](https://github.com/apache/pinot/issues/12888)

With publishNotReadyAddresses set to true in service resource, kubernetes will publish the DNS address irrespective of whether the pod is in Ready state or not. Then the sole responsibility of when to start making requests to the newly restarted server/broker lies on Pinot. 

The following error will be fixed with this PR

````
Unable to resolve host pinot-server-2.pinot-server-headless.pinot-dev.svc.cluster.local
or 
Unable to resolve host pinot-broker-2.pinot-server-headless.pinot-dev.svc.cluster.local
````

### CODE REVIEW

**Comment by abhioncbr** [comment]
Does it mean that with this change, pods are going to get the request before they get into Ready Stage?

**Comment by piby180** [comment]
1. Currently, Pinot assumes that the server is online as soon as all the segments have been made online on zookeeper and it starts to send the requests to the newly restarted server immediately.
2. Kubernetes however marks the server pod as "Ready" only when the first readiness probe is succeeded. 
3.  Since pinot makes request to headless service, in order for the request to resolve, the DNS for each server is needed to be resolved. So all the queries made to Pinot between "the time the server is online as per zookeeper" and "the time kubernetes mark it as Ready" fails with error.
````
Unable to resolve host pinot-server-2.pinot-server-headless.pinot-dev.svc.cluster.local
or 
Unable to resolve host pinot-broker-2.pinot-server-headless.pinot-dev.svc.cluster.local
````
4.  This causes downtime in Pinot whenver a pod (server or broker) is restarted.  This can be seconds or minutes depending on how the probes are configured. 

With this change, the DNS for each server will always resolve irrespective of whether the pod is in "Ready" state or not. Pinot will however only make requests to the server when it is online as per Zookeeper.

**Comment by abhioncbr** [comment]
> 1. Currently, Pinot assumes that the server is online as soon as all the segments have been made online on zookeeper and it starts to send the requests to the newly restarted server immediately.
> 2. Kubernetes however marks the server pod as "Ready" only when the first readiness probe is succeeded.
> 3. Since pinot makes request to headless service, in order for the request to resolve, the DNS for each server is needed to be resolved. So all the queries made to Pinot between "the time the server is online as per zookeeper" and "the time kubernetes mark it as Ready" fails with error.
> 
> ```
> Unable to resolve host pinot-server-2.pinot-server-headless.pinot-dev.svc.cluster.local
> or 
> Unable to resolve host pinot-broker-2.pinot-server-headless.pinot-dev.svc.cluster.local
> ```
> 
> 4. This causes downtime in Pinot whenver a pod (server or broker) is restarted.  This can be seconds or minutes depending on how the probes are configured.
> 
> With this change, the DNS for each server will always resolve irrespective of whether the pod is in "Ready" state or not. Pinot will however only make requests to the server when it is online as per Zookeeper.

Thanks for the explanation. I just checked, and the zookeeper is doing the same for the headless service.
Question: Should it be added only in headless services or in all services?

**Comment by piby180** [comment]
> > 1. Currently, Pinot assumes that the server is online as soon as all the segments have been made online on zookeeper and it starts to send the requests to the newly restarted server immediately.
> > 2. Kubernetes however marks the server pod as "Ready" only when the first readiness probe is succeeded.
> > 3. Since pinot makes request to headless service, in order for the request to resolve, the DNS for each server is needed to be resolved. So all the queries made to Pinot between "the time the server is online as per zookeeper" and "the time kubernetes mark it as Ready" fails with error.
> > 
> > ```
> > Unable to resolve host pinot-server-2.pinot-server-headless.pinot-dev.svc.cluster.local
> > or 
> > Unable to resolve host pinot-broker-2.pinot-server-headless.pinot-dev.svc.cluster.local
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > 4. This causes downtime in Pinot whenver a pod (server or broker) is restarted.  This can be seconds or minutes depending on how the probes are configured.
> > 
> > With this change, the DNS for each server will always resolve irrespective of whether the pod is in "Ready" state or not. Pinot will however only make requests to the server when it is online as per Zookeeper.
> 
> Thanks for the explanation. I just checked, and the zookeeper is doing the same for the headless service. Question: Should it be added only in headless services or in all services?

Change to just headless services is sufficient. This has been tested on our cluster. 
The "normal" service endpoints are being used by ingress resource. 

If you agree, I can remove this change from "headful" services.

**Comment by abhioncbr** [comment]
> > > 1. Currently, Pinot assumes that the server is online as soon as all the segments have been made online on zookeeper and it starts to send the requests to the newly restarted server immediately.
> > > 2. Kubernetes however marks the server pod as "Ready" only when the first readiness probe is succeeded.
> > > 3. Since pinot makes request to headless service, in order for the request to resolve, the DNS for each server is needed to be resolved. So all the queries made to Pinot between "the time the server is online as per zookeeper" and "the time kubernetes mark it as Ready" fails with error.
> > > 
> > > ```
> > > Unable to resolve host pinot-server-2.pinot-server-headless.pinot-dev.svc.cluster.local
> > > or 
> > > Unable to resolve host pinot-broker-2.pinot-server-headless.pinot-dev.svc.cluster.local
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > 
> > > 4. This causes downtime in Pinot whenver a pod (server or broker) is restarted.  This can be seconds or minutes depending on how the probes are configured.
> > > 
> > > With this change, the DNS for each server will always resolve irrespective of whether the pod is in "Ready" state or not. Pinot will however only make requests to the server when it is online as per Zookeeper.
> > 
> > 
> > Thanks for the explanation. I just checked, and the zookeeper is doing the same for the headless service. Question: Should it be added only in headless services or in all services?
> 
> Change to just headless services is sufficient. This has been tested on our cluster. The "normal" service endpoints are being used by ingress resource.
> 
> If you agree, I can remove this change from "headful" services.

In my opinion, it should be only in headless svc, but I suggest we wait for others' input as well.

**Review by AustinRP-Primer — APPROVED** [review]

**Review by zhtaoxiang — APPROVED** [review]
LGTM

**Review by xiangfu0 — APPROVED** [review]

File: helm/pinot/templates/broker/service-external.yaml [hunk 1]
```diff
@@ -29,6 +29,7 @@ metadata:
     {{- include "pinot.brokerLabels" . | nindent 4 }}
 spec:
   type: {{ .Values.broker.external.type }}
+  publishNotReadyAddresses: true
```
Comment (xiangfu0) [inline helm/pinot/templates/broker/service-external.yaml:32]: let's make default external to false here.

**Comment by xiangfu0** [comment]
I think we should only make default value for service-headless to true, all others to false.

**Comment by xiangfu0** [comment]
Pinot handles all the internal readiness for routing request to servers, so it's ok to expose all the server ips.
For controller and broker, they all start up pretty fast, so it's ok to wait until health check finishing to mark them online.

File: helm/pinot/templates/broker/service.yaml [hunk 2]
```diff
@@ -27,6 +27,7 @@ metadata:
     {{- include "pinot.brokerLabels" . | nindent 4 }}
 spec:
   type: {{ .Values.broker.service.type }}
+  publishNotReadyAddresses: true
```
Comment (xiangfu0) [inline helm/pinot/templates/broker/service.yaml:30]: default to false

File: helm/pinot/templates/controller/service-external.yaml [hunk 3]
```diff
@@ -29,6 +29,7 @@ metadata:
     {{- include "pinot.controllerLabels" . | nindent 4 }}
 spec:
   type: {{ .Values.controller.external.type }}
+  publishNotReadyAddresses: true
```
Comment (xiangfu0) [inline helm/pinot/templates/controller/service-external.yaml:32]: default to false

File: helm/pinot/templates/controller/service.yaml [hunk 4]
```diff
@@ -27,6 +27,7 @@ metadata:
     {{- include "pinot.controllerLabels" . | nindent 4 }}
 spec:
   type: {{ .Values.controller.service.type }}
+  publishNotReadyAddresses: true
```
Comment (xiangfu0) [inline helm/pinot/templates/controller/service.yaml:30]: default to false

Comment (piby180) [reply helm/pinot/templates/controller/service.yaml]: done

File: helm/pinot/templates/controller/service-external.yaml (hunk 3 above)
Comment (piby180) [reply helm/pinot/templates/controller/service-external.yaml]: done

File: helm/pinot/templates/broker/service.yaml (hunk 2 above)
Comment (piby180) [reply helm/pinot/templates/broker/service.yaml]: done

File: helm/pinot/templates/broker/service-external.yaml (hunk 1 above)
Comment (piby180) [reply helm/pinot/templates/broker/service-external.yaml]: done

**Comment by piby180** [comment]
@xiangfu0  Can we merge this?

**Comment by davidballano** [comment]
Yes please, can we merge this, otherwise we might have to fork the chart.

**Pull request merged**